src/graph.rs

```diff
@@ -678,6 +679,162 @@ pub struct BuildOptions<'a> {
   pub reporter: Option<&'a dyn Reporter>,
 }
 
+#[derive(Debug, Copy, Clone)]
+pub enum ModuleEntryRef<'a> {
+  Module(&'a Module),
+  Err(&'a ModuleGraphError),
+  Redirect(&'a ModuleSpecifier),
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct WalkOptions {
+  pub follow_dynamic: bool,
+  pub follow_type_only: bool,
+  pub check_js: bool,
+}
+
+impl Default for WalkOptions {
+  fn default() -> Self {
+    Self {
+      follow_dynamic: true,
+      follow_type_only: true,
+      check_js: true,
+    }
+  }
+}
+
+pub struct ModuleEntryIterator<'a> {
+  graph: &'a ModuleGraph,
+  seen: HashSet<&'a ModuleSpecifier>,
+  visiting: VecDeque<&'a ModuleSpecifier>,
+  follow_dynamic: bool,
+  follow_type_only: bool,
+  check_js: bool,
+}
+
+impl<'a> ModuleEntryIterator<'a> {
+  fn new(
+    graph: &'a ModuleGraph,
+    roots: &[&'a ModuleSpecifier],
+    options: WalkOptions,
+  ) -> Self {
+    let mut seen = HashSet::<&'a ModuleSpecifier>::with_capacity(
+      graph.roots.len() + graph.redirects.len(),
+    );
+    let mut visiting = VecDeque::<&'a ModuleSpecifier>::new();
+    for root in roots {
+      seen.insert(root);
+      visiting.push_back(root);
+    }
+    for (_, dep) in graph.imports.values().flat_map(|i| &i.dependencies) {
+      let mut resolutions = Vec::with_capacity(2);
+      resolutions.push(&dep.maybe_code);
+      if options.follow_type_only {
+        resolutions.push(&dep.maybe_type);
+      }
+      #[allow(clippy::manual_flatten)]
+      for resolved in resolutions {
+        if let Resolved::Ok { specifier, .. } = resolved {
+          if !seen.contains(specifier) {
+            seen.insert(specifier);
+            visiting.push_front(specifier);
+          }
+        }
+      }
+    }
+    Self {
+      graph,
+      seen,
+      visiting,
+      follow_dynamic: options.follow_dynamic,
+      follow_type_only: options.follow_type_only,
+      check_js: options.check_js,
+    }
+  }
+}
+
+impl<'a> Iterator for ModuleEntryIterator<'a> {
+  type Item = (&'a ModuleSpecifier, ModuleEntryRef<'a>);
+
+  fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This body of this is essentially copy and paste from the CLI, but adapted to work as an iterator.